### ISSUE TITLE
Storage writer java.sql,Exception MOETEREGISTRERING

### ISSUE THREAD

**Issue by solfeggietto** [documaster/noark-extraction-validator#28]
Attached files describing the Issue.
Version: 0.2.0_SNAPSHOT

Exception in thread "Thread-2" com.documaster.validator.exceptions.StorageException: 
Storage writer encountered an unexpected error
at com.documaster.validator.storage.core.StorageWriter.run(StorageWriter.java:77)

Caused by: java.sql.SQLException: Column not found: MOETEREGISTRERING in statement
[INSERT INTO arkivstruktur.registrering ( _id, _ref_mappe, _parent_id, ref_klasse, _ref_arkivdel, _dtype, systemid, opprettetav, referansearkivdel, registreringsid, tittel, offentligtittel, beskrivelse, forfatter, dokumentmedium, moeteregistreringstype, motesakstype, moeteregistreringsstatus, administrativenhet, saksbehandler, moeteregistrering ) VALUES ( ?, etc 22 times 23)

=> 23 Fields, 22 Values ? in log window Exception
Perhaps the 23rd Field MOETEREGISRERING then just is one short of Values passed in the SQL call?

![validering-feilmelding1](https://cloud.githubusercontent.com/assets/622856/23950055/06be2d5a-098a-11e7-9def-9d27cfcbc6c3.png)
[rapport_bat.txt](https://github.com/documaster/noark-extraction-validator/files/844580/rapport_bat.txt)
[Erfaringer.txt](https://github.com/documaster/noark-extraction-validator/files/844581/Erfaringer.txt)
[Arkiv Sola kommune 2017-03-14 10-38-13.xlsx](https://github.com/documaster/noark-extraction-validator/files/844583/Arkiv.Sola.kommune.2017-03-14.10-38-13.xlsx)
[dn5v_log.txt](https://github.com/documaster/noark-extraction-validator/files/844585/dn5v_log.txt)

**Comment by ivaylomitrev** [comment]
**Preamble**
In the attached screenshot there are actually 22 fields and 22 `?` placeholders. @solfeggietto, you have missed `opprettetdato` when rewriting the SQL query. I have described the actual reason for the issue below. Also, please take a look at my request at the end of the comment.

**Reproduction scenario:**
I have been able to reproduce the exception locally. What I needed to do was: 
1. manually modify an `arkivstruktur.xml` file containing a `moeteregistrering` and introduce a `moeteregistrering` field in it such that:
```
<registrering xsi:type="moeteregistrering">
    <systemID>Test</systemID>
    <opprettetDato>2017-01-01T00:00:01</opprettetDato>
    <opprettetAv>Test</opprettetAv>
    <arkivertDato>2017-01-01T00:00:01</arkivertDato>
    <arkivertAv>Test</arkivertAv>
    <dokumentbeskrivelse>
        <systemID>systemID</systemID>
        <dokumenttype>[empty]</dokumenttype>
        <dokumentstatus>Dokumentet er ferdigstilt</dokumentstatus>
        <tittel>test file.pdfa</tittel>
        <opprettetDato>2017-01-01T00:00:00</opprettetDato>
        <opprettetAv>Test</opprettetAv>
        <dokumentmedium>Elektronisk arkiv</dokumentmedium>
        <tilknyttetRegistreringSom>Hoveddokument</tilknyttetRegistreringSom>
        <dokumentnummer>1</dokumentnummer>
        <tilknyttetDato>2017-01-01T00:00:00</tilknyttetDato>
        <tilknyttetAv>Test</tilknyttetAv>
        <dokumentobjekt>
            <versjonsnummer>1</versjonsnummer>
            <variantformat>Produksjonsformat</variantformat>
            <format>[empty]</format>
            <opprettetDato>2017-01-01T00:00:00</opprettetDato>
            <opprettetAv>Test</opprettetAv>
            <referanseDokumentfil>dokumenter/test.txt</referanseDokumentfil>
            <sjekksum>9a30a503b2862c51c3c5acd7fbce2f1f784cf4658ccf8e87d5023a90c21c0714</sjekksum>
            <sjekksumAlgoritme>SHA-256</sjekksumAlgoritme>
            <filstoerrelse>9</filstoerrelse>
        </dokumentobjekt>
    </dokumentbeskrivelse>
    <registreringsID>2017/1</registreringsID>
    <tittel>Test</tittel>
    <offentligTittel>Test</offentligTittel>
    <dokumentmedium>Elektronisk arkiv</dokumentmedium>
    <moeteregistreringstype>Inngående dokument</moeteregistreringstype>
    <moetesakstype>Test</moetesakstype>
    <moeteregistreringsstatus>Arkivert</moeteregistreringsstatus>
    <administrativEnhet>Test</administrativEnhet>
    <saksbehandler>Test</saksbehandler>
    <moeteregistrering>Test</moeteregistrering>
</registrering>
```
(The above is just an illustration. Notice the last field of the `moeteregistrering` which should not be there according to the Noark 5 standard).

2. Run the validation with the `-ignore-non-compliant-xml` flag

**Cause of the issue**
The reason for the issue is the addition of an unexpected column in an entity of type `registrering`: `moeteregistrering`. Generally this is a critical integrity violation of the standard which would force the validator to stop execution with an exception message: `Validation has been stopped due to unrecoverable extraction integrity violation.`. In such cases the generated reports would indicate the following issue:
```
cvc-complex-type.2.4.a: Invalid content was found starting with element 'moeteregistrering'. One of '{"http://www.arkivverket.no/standarder/noark5/arkivstruktur":referanseTilMoeteregistrering, "http://www.arkivverket.no/standarder/noark5/arkivstruktur":referanseFraMoeteregistrering}' is expected.
```
If the validation is run with the `ignore-non-compliant-xml` flag, however, execution will continue and will lead to the issue described by Torbjørn. **As suggested by the flag's description, the validation results should not be trusted and the validation may result in unpredictable outcomes when it is used.**

**Next steps**
Generally, the issue should be reproducible for any extraction package that contains entities whose children are not defined in the standard's schemas. The reason for not handling such cases initially was that this is an _unrecoverable integrity violation_ and it may lead to a ton of unpredictable results during the actual validation phase. Considering the latest development of the validator towards better handling of invalid extraction packages, however, I will address this issue. I intend to disallow the insertion of unrecognized fields in the relational database that backs the validation. This means that information in unrecognized entities/fields will be completely ignored by the validator. I see no better workaround, because there is not a good way to determine if their content would be useful or not.

@douzounov, I am tagging you so that you are aware of the issue and its suggested fix.

@solfeggietto, I have just one request: Can you confirm that the assumptions made above match your scenario? That is, does `arkivstruktur.xml` contain a `moeteregistrering` that has a `moeteregistrering` child and that you have run the validation using the `-ignore-non-compliant-xml` flag?

**Comment by ivaylomitrev** [comment]
The issue has been fixed and will be available in release 0.2.0.
